synapse/storage/databases/main/room.py

```diff
@@ -1569,6 +1639,69 @@ def _background_populate_rooms_creator_column_txn(
 
         return batch_size
 
+    async def _background_add_room_type_column(
+        self, progress: JsonDict, batch_size: int
+    ) -> int:
+        """Background update to go and add room_type information to `room_stats_state`
+        table from `event_json` table.
+        """
+
+        last_room_id = progress.get("room_id", "")
+
+        def _background_add_room_type_column_txn(
+            txn: LoggingTransaction,
+        ) -> bool:
+            sql = """
+                SELECT state.room_id, json FROM event_json
+                INNER JOIN current_state_events AS state USING (event_id)
+                WHERE state.room_id > ? AND type = 'm.room.create'
+                ORDER BY state.room_id
+                LIMIT ?
+            """
+
+            txn.execute(sql, (last_room_id, batch_size))
+            room_id_to_create_event_results = txn.fetchall()
+
+            new_last_room_id = ""
+            for room_id, event_json in room_id_to_create_event_results:
+                event_dict = db_to_json(event_json)
+
+                room_type = event_dict.get("content", {}).get(
+                    EventContentFields.ROOM_TYPE, None
+                )
+                if room_type:
+                    self.db_pool.simple_update_txn(
+                        txn,
+                        table="room_stats_state",
+                        keyvalues={"room_id": room_id},
+                        updatevalues={"room_type": room_type},
+                    )
+
+                new_last_room_id = room_id
+
+            if new_last_room_id == "":
+                return True
```

Comment:
nit: I'm in favour of using `None` over empty string to mean "no last room".

```suggestion
            new_last_room_id = None
            for room_id, event_json in room_id_to_create_event_results:
                event_dict = db_to_json(event_json)

                room_type = event_dict.get("content", {}).get(
                    EventContentFields.ROOM_TYPE, None
                )
                if room_type:
                    self.db_pool.simple_update_txn(
                        txn,
                        table="room_stats_state",
                        keyvalues={"room_id": room_id},
                        updatevalues={"room_type": room_type},
                    )

                new_last_room_id = room_id

            if new_last_room_id is None:
                return True
```